test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java

```diff
@@ -277,6 +277,10 @@ public void testClass() {
     assertNoWarning("class A { f() { return 1729; } }");
   }
 
+  public void testRedeclareClassName() {
+    assertNoWarning("var Clazz = class Foo {}; var Foo = 3;");
```

Comment:
I originally encountered this with a custom Promise implementation. This bug is easy to work around, but the behavior is inconsistent and surprising considering the compiler properly handles scoping on [function names](http://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520Clazz%2520%253D%2520class%2520Foo%2520%257B%257D%253B%250Aclass%2520Foo%2520%257B%257D%250A%250Aclass%2520Bar%2520%257B%257D%250Avar%2520Baz%2520%253D%2520class%2520Bar%2520%257B%257D%250A%250Avar%2520one%2520%253D%2520function%2520two()%2520%257B%257D%253B%250Afunction%2520two()%2520%257B%257D%250A%250Afunction%2520three()%2520%257B%257D%250Avar%2520four%2520%253D%2520function%2520three()%2520%257B%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1)